dataset/src/main/scala/frameless/macros/ColumnMacros.scala

```diff
@@ -0,0 +1,377 @@
+              errs match {
+                case (tree, err) :: Nil => fail(tree)(s": $err")
+                case multi =>
+                  multi foreach {
+                    case (tree, err) => c.error(tree.pos, err)
+                  }
+                  fail(body)(": multiple errors composing projection")
+              }
+          }
+
+          case other =>
+            val o = other
+            fail(o)(": could not compose projection")
+        }
+        val df = q"${c.prefix}.dataset.toDF().select(..$selectCols)"
+        val ds = q"$df.as[$B]($TypedExpressionEncoder.apply[$B])"
+        val typedDs = q"$TypedDataset.create[$B]($ds)($encoder)"
+
+        typedDs
+    }
+
+
+    c.typecheck(result)
+  }
+
+  sealed trait Expr {
+    def tpe: Type
+  }
+
+  object Expr {
+    // zip down to the leaves and construct an expression that gives us a Column
+    // this is some kind of hylo-mylo-bananafanafofylo-morphism.
+    def refold(expr: Expr): Tree = expr match {
+      case LiteralExpr(tpe, tree) => q"org.apache.spark.sql.functions.lit($tree)"
+      case Column(tpe, selectorStr) =>
+        q"new $ColumnName($selectorStr)"
+      // $COVERAGE-OFF$ - Functions can't currently be used (and currently give an error before this is reached)
+      case FunctionApplication(tpe, src, fn, args) =>
+        c.abort(c.enclosingPosition, "Functions not yet supported in selectExpr")
+      // $COVERAGE-ON$
+      case BinaryOperator(tpe, lhs, op, rhs) =>
+        val lhsTree = refold(lhs)
+        val rhsTree = refold(rhs)
+        Apply(Select(lhsTree, op), List(rhsTree))
+      case UnaryRAOperator(tpe, rhs, op) =>
+        val rhsTree = refold(rhs)
+        Select(rhsTree, op)
+      case Construct(tpe, args) =>
+        val argTrees = toNamedColumns(args)
+        q"org.apache.spark.sql.functions.struct(..$argTrees)"
+      case TrustMeBro(tpe, tree) => tree
+    }
+
+    def toNamedColumns(exprMap: Map[String, Expr]): List[Tree] = exprMap.toList.map {
+      case (colName, expr) =>
+        val colTree = refold(expr)
+        q"$colTree.as($colName)"
+    }
+  }
+
+  case class Construct(tpe: Type, args: Map[String, Expr]) extends Expr                  // Construct tuple or case class
+  case class UnaryRAOperator(tpe: Type, rhs: Expr, op: TermName) extends Expr            // Unary right-associative op
+  case class BinaryOperator(tpe: Type, lhs: Expr, op: TermName, rhs: Expr) extends Expr  // Binary op
+  case class FunctionApplication(tpe: Type, src: Tree, fn: TermName, args: List[Expr]) extends Expr   // SQL function
+  case class Column(tpe: Type, selectorStr: String) extends Expr                         // Column of this dataset
+  case class LiteralExpr(tpe: Type, tree: Tree) extends Expr                             // Literal expression
+
+  // manually added tree (temporary until functions can be implemented; needed for operator substitution)
+  case class TrustMeBro(tpe: Type, tree: Tree) extends Expr
```

Comment:
Can get rid of this once we figure out the best way to support function application